pkg/controller/endpointslice/endpointslice_tracker.go

```diff
@@ -101,6 +100,51 @@ func (est *endpointSliceTracker) StaleSlices(service *v1.Service, endpointSlices
 	return false
 }
 
+// MissingSlices returns true if this tracker is tracking more EndpointSlices
+// for the provided Service than have been provided to this function. If this
+// function returns true, it can be used as an indication that the controller
+// should fall back to a full API list instead of relying on the cache.
+func (est *endpointSliceTracker) MissingSlices(namespace, name string, endpointSlices []*discovery.EndpointSlice) bool {
+	est.lock.Lock()
+	defer est.lock.Unlock()
+
+	nn := types.NamespacedName{Name: name, Namespace: namespace}
+	gfs, ok := est.generationsByService[nn]
+	if !ok {
+		return false
+	}
+	providedSlices := make(map[types.UID]bool, len(endpointSlices))
+	for _, endpointSlice := range endpointSlices {
+		providedSlices[endpointSlice.UID] = true
+	}
+	for trackedUID, gen := range gfs {
+		if gen == deletionExpected {
+			continue
+		}
+		_, ok := providedSlices[trackedUID]
+		if !ok {
+			return true
+		}
+	}
+	return false
+}
+
+// Reset updates the tracker to contain only the provided slices for the
+// specified Service.
+func (est *endpointSliceTracker) Reset(namespace, name string, endpointSlices []*discovery.EndpointSlice) {
```

Comment:
I'm asking because another option is to do those operations in the main loop and not having to add a new method to tracker, it is growing too much in size and complexity ...

Reply:
scratch my comment, that will not work because we are not holding the lock , and there will not be an unique transaction, sorry